src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
+                                    // Return None and create another.
+                                    if e_conn.at_max_streamcount() {
+                                        debug!("found existing conn for key {:#?}, but streamcount is maxed", workload_key);
+                                        break None;
+                                    }
+                                    break Some(e_conn);
+                                }
+                            }
+                        }
+                        Err(_) => {
+                            return Err(Error::WorkloadHBONEPoolDraining);
+                        }
+                    }
+                }
+            }
+        };
+
+        // If we get here, it means the following are true:
+        // 1. At one point, there was a preexisting conn in the pool for this key.
+        // 2. When we checked, we got nothing for that key.
+        // 3. We could not get the exclusive inner writelock to add a new one
+        // 4. Someone else got the exclusive inner writelock, and is adding a new one
+        // 5. We waited until we got that connection for our key, that someone else added.
+        //
+        // So now we are down to 2 options:
+        //
+        // 1. We got an existing connection, it's usable, we use it.
+        // 2. We got an existing connection, it's not usable, we start another.
+        //
+        // Note that for (2) we do not really need to take the inner writelock again for this key -
+        // if we checked out a conn inserted by someone else that we cannot use,
+        // everyone else will implicitly be waiting for us to check one back in,
+        // since they are in their own loops.
+        match got_conn {
+            // After waiting, we found an available conn we can use, so no need to start another.
+            // Clone the underlying client, return a copy, and put the other back in the pool.
+            Some(f_conn) => {
+                self.state.checkin_conn(f_conn.clone(), pool_key.clone());
+                Ok(f_conn)
+            }
+
+            // After waiting, we found an available conn, but for whatever reason couldn't use it.
+            // (streamcount maxed, etc)
+            // Start a new one, clone the underlying client, return a copy, and put the other back in the pool.
+            None => {
+                debug!("spawning new conn for key {:#?} to replace", workload_key);
+                let pool_conn = self.spawn_new_pool_conn(workload_key.clone()).await;
+                let r_conn = ConnClient {
+                    sender: pool_conn?,
+                    stream_count: Arc::new(AtomicU16::new(0)),
+                    stream_count_max: self.max_streamcount,
+                    wl_key: workload_key.clone(),
+                };
+                self.state.checkin_conn(r_conn.clone(), pool_key.clone());
+                Ok(r_conn)
+            }
```

Comment:
If we are under hash collision AND we happened to be at max streams we can now potentially pollute the pool with a connection to a different dst I think. Even MORE unlikely but I think technically possible

Reply:
Ah, that's a good call. Yep. TY!